components/prism-core.js

```diff
@@ -189,25 +208,17 @@ var _ = {
 
 	highlightElement: function(element, async, callback) {
 		// Find language
-		var language = 'none', grammar, parent = element;
-
-		while (parent && !lang.test(parent.className)) {
-			parent = parent.parentNode;
-		}
-
-		if (parent) {
-			language = (parent.className.match(lang) || [,'none'])[1].toLowerCase();
-			grammar = _.languages[language];
-		}
+		var language = getLanguage(element);
+		var grammar = _.languages[language];
 
 		// Set language on the element, if not present
 		element.className = element.className.replace(lang, '').replace(/\s+/g, ' ') + ' language-' + language;
 
 		if (element.parentNode) {
 			// Set language on the parent, for styling
-			parent = element.parentNode;
+			var parent = element.parentNode;
 
-			if (/pre/i.test(parent.nodeName)) {
+			if (/^pre$/i.test(parent.nodeName)) {
```

Comment:
```suggestion
			if (parent.nodeName.toLowerCase() === 'pre') {
```
Wouldn't something like this be more efficient?

Reply:
Good question. Transforming & allocating a string + compare or traversing what is effectively a DFA.
(Inline patterns are cached, so that's not an issue.)

Reply:
Ok. I benchmarked it and the regex approach is slightly slower.